bug(forensics): audit trail hash chain drops expected entries under regression/race suite

## Summary
The full regression gate currently fails in `internal/harness` because `TestAuditTrail_HashChainValid` sometimes sees only 2 audit entries when it expects at least 3 (`run.started`, `audit.action`, `run.completed`). This leaves the audit trail subsystem in a not-fully-stable state even though the main non-race suite passes.

## Impact
- `./scripts/test-regression.sh` is currently red on a core backend reliability path.
- The append-only audit trail is one of the highest-trust forensic surfaces in the repo; nondeterministic missing entries directly undermine that value.
- The repo cannot claim a clean regression gate while this bug remains.

## Reproduction
From the repo root:

```bash
./scripts/test-regression.sh
```

Observed failure during the race/regression pass:

```text
--- FAIL: TestAuditTrail_HashChainValid (0.01s)
    runner_audittrail_test.go:213: expected >= 3 audit entries (started + action + completed), got 2
FAIL
FAIL    go-agent-harness/internal/harness
```

## Current Behavior
- The test expectation is in `internal/harness/runner_audittrail_test.go:200-220`.
- Audit action emission/write lives in `internal/harness/runner.go:1789-1802`.
- Audit writer creation/ownership lives in `internal/harness/runner.go:386-397` and `internal/harness/runner.go:4364-4409`.

## Expected Behavior
- Every state-modifying tool call that qualifies for audit logging should result in a persisted `audit.action` entry before the audit writer is closed.
- `TestAuditTrail_HashChainValid` should pass reliably under the full regression suite, including the race pass.

## Suspected Root Cause
One of the following is likely happening, but this issue should confirm the exact cause rather than assume it:
- `audit.action` is emitted but not persisted before writer close/drain.
- The state-modifying-tool classification is inconsistent for the test path.
- Audit write ordering/lifecycle has a race with terminal run completion.

## Proposed Fix Direction
1. Reproduce the failure deterministically (loop the test and/or isolate the race-path execution).
2. Trace the exact audit lifecycle for `run.started`, `audit.action`, and terminal completion.
3. Fix the lifecycle/order bug so the action entry is durably written before close.
4. Add a stronger regression test if needed (for example repeated-run coverage or explicit synchronization assertions) so this cannot silently regress again.

## Acceptance Criteria
- `TestAuditTrail_HashChainValid` passes reliably on repeated local runs.
- `./scripts/test-regression.sh` passes cleanly after the fix.
- The final audit log contains the expected started/action/completed chain for the reproducer scenario.

## Test Plan / Regression Coverage
- Re-run the isolated failing test in a loop.
- Re-run `go test ./internal/harness -race`.
- Re-run `./scripts/test-regression.sh`.

## Related Issues / References
- Original audit trail feature: #214
- Relevant files:
  - `internal/harness/runner.go`
  - `internal/harness/runner_audittrail_test.go`
  - `internal/forensics/audittrail/*`

--- a/internal/harness/runner.go
+++ b/internal/harness/runner.go
@@ -2641,8 +2641,6 @@ func (r *Runner) drainSteering(runID string, messages *[]Message) {
 }
 
 func (r *Runner) completeRun(runID, output string) {
-	r.setStatus(runID, RunStatusCompleted, output, "")
-
 	// Clean up deferred tool activations for this run
 	r.activations.Cleanup(runID)
 
@@ -2704,13 +2702,6 @@ func (r *Runner) completeRun(runID, output string) {
 		r.mu.RUnlock()
 	}
 
-	usageTotals, costTotals := r.accountingTotals(runID)
-	r.emit(runID, EventRunCompleted, map[string]any{
-		"output":       output,
-		"usage_totals": usageTotals,
-		"cost_totals":  costTotals,
-	})
-
 	// Audit trail: write run.completed and close the writer.
 	if r.config.AuditTrailEnabled {
 		r.writeAudit(runID, audittrail.AuditRecord{
@@ -2720,6 +2711,15 @@ func (r *Runner) completeRun(runID, output string) {
 		})
 		r.closeAuditWriter(runID)
 	}
+
+	r.setStatus(runID, RunStatusCompleted, output, "")
+
+	usageTotals, costTotals := r.accountingTotals(runID)
+	r.emit(runID, EventRunCompleted, map[string]any{
+		"output":       output,
+		"usage_totals": usageTotals,
+		"cost_totals":  costTotals,
+	})
 }
 
 // closeScopedMCP closes the per-run scoped MCP registry if one was configured
@@ -2855,7 +2855,6 @@ func (r *Runner) failRun(runID string, err error) {
 	if err == nil {
 		err = errors.New("run failed")
 	}
-	r.setStatus(runID, RunStatusFailed, "", err.Error())
 
 	// Clean up deferred tool activations for this run
 	r.activations.Cleanup(runID)
@@ -2882,13 +2881,6 @@ func (r *Runner) failRun(runID string, err error) {
 		}
 	}
 
-	usageTotals, costTotals := r.accountingTotals(runID)
-	r.emit(runID, EventRunFailed, map[string]any{
-		"error":        err.Error(),
-		"usage_totals": usageTotals,
-		"cost_totals":  costTotals,
-	})
-
 	// Audit trail: write run.failed and close the writer.
 	if r.config.AuditTrailEnabled {
 		r.writeAudit(runID, audittrail.AuditRecord{
@@ -2898,6 +2890,15 @@ func (r *Runner) failRun(runID string, err error) {
 		})
 		r.closeAuditWriter(runID)
 	}
+
+	r.setStatus(runID, RunStatusFailed, "", err.Error())
+
+	usageTotals, costTotals := r.accountingTotals(runID)
+	r.emit(runID, EventRunFailed, map[string]any{
+		"error":        err.Error(),
+		"usage_totals": usageTotals,
+		"cost_totals":  costTotals,
+	})
 }
 
 // failRunMaxSteps is a specialisation of failRun used when the step loop
@@ -2906,7 +2907,6 @@ func (r *Runner) failRun(runID string, err error) {
 // this terminal state from other failures without parsing the error string.
 func (r *Runner) failRunMaxSteps(runID string, maxSteps int) {
 	err := fmt.Errorf("max steps (%d) reached", maxSteps)
-	r.setStatus(runID, RunStatusFailed, "", err.Error())
 
 	// Clean up deferred tool activations for this run
 	r.activations.Cleanup(runID)
@@ -2916,15 +2916,6 @@ func (r *Runner) failRunMaxSteps(runID string, maxSteps int) {
 
 	// Clean up per-run MCP servers
 	r.closeScopedMCP(runID)
-	usageTotals, costTotals := r.accountingTotals(runID)
-	r.emit(runID, EventRunFailed, map[string]any{
-		"error":        err.Error(),
-		"reason":       "max_steps_reached",
-		"max_steps":    maxSteps,
-		"usage_totals": usageTotals,
-		"cost_totals":  costTotals,
-	})
-
 	// Audit trail: write run.failed and close the writer.
 	if r.config.AuditTrailEnabled {
 		r.writeAudit(runID, audittrail.AuditRecord{
@@ -2938,6 +2929,17 @@ func (r *Runner) failRunMaxSteps(runID string, maxSteps int) {
 		})
 		r.closeAuditWriter(runID)
 	}
+
+	r.setStatus(runID, RunStatusFailed, "", err.Error())
+
+	usageTotals, costTotals := r.accountingTotals(runID)
+	r.emit(runID, EventRunFailed, map[string]any{
+		"error":        err.Error(),
+		"reason":       "max_steps_reached",
+		"max_steps":    maxSteps,
+		"usage_totals": usageTotals,
+		"cost_totals":  costTotals,
+	})
 }
 
 func (r *Runner) recordAccounting(runID string, result CompletionResult, step int) map[string]any {

--- a/internal/harness/runner_audittrail_test.go
+++ b/internal/harness/runner_audittrail_test.go
@@ -7,12 +7,23 @@ import (
 	"errors"
 	"os"
 	"path/filepath"
+	"syscall"
 	"testing"
+	"time"
 
 	"go-agent-harness/internal/forensics/audittrail"
 )
 
 var errAuditTest = errors.New("audit test provider error")
+
+type blockingAuditProvider struct {
+	release <-chan struct{}
+}
+
+func (p *blockingAuditProvider) Complete(_ context.Context, _ CompletionRequest) (CompletionResult, error) {
+	<-p.release
+	return CompletionResult{Content: "done"}, nil
+}
 
 // readAuditEntries reads all entries from an audit.jsonl file.
 func readAuditEntries(t *testing.T, path string) []audittrail.AuditEntry {
@@ -222,6 +233,87 @@ func TestAuditTrail_HashChainValid(t *testing.T) {
 			t.Errorf("chain broken at %d: PrevHash=%q != entries[%d].EntryHash=%q",
 				i, entries[i].PrevHash, i-1, entries[i-1].EntryHash)
 		}
+	}
+}
+
+func TestAuditTrail_RunCompletedStatusWaitsForAuditPersistence(t *testing.T) {
+	t.Parallel()
+
+	dir := t.TempDir()
+	release := make(chan struct{})
+	runner := NewRunner(&blockingAuditProvider{release: release}, NewRegistry(), RunnerConfig{
+		DefaultModel:        "test-model",
+		DefaultSystemPrompt: "You are helpful.",
+		MaxSteps:            1,
+		RolloutDir:          dir,
+		AuditTrailEnabled:   true,
+	})
+
+	run, err := runner.StartRun(RunRequest{Prompt: "complete once"})
+	if err != nil {
+		t.Fatalf("StartRun: %v", err)
+	}
+
+	auditPath := auditLogPath(dir)
+	deadline := time.Now().Add(2 * time.Second)
+	for {
+		entries := readAuditEntries(t, auditPath)
+		if len(entries) >= 1 && entries[0].EventType == string(EventRunStarted) {
+			break
+		}
+		if time.Now().After(deadline) {
+			t.Fatalf("timed out waiting for run.started audit entry")
+		}
+		time.Sleep(10 * time.Millisecond)
+	}
+
+	lockFile, err := os.OpenFile(auditPath, os.O_RDWR, 0)
+	if err != nil {
+		t.Fatalf("OpenFile(%s): %v", auditPath, err)
+	}
+	locked := false
+	defer func() {
+		if lockFile == nil {
+			return
+		}
+		if locked {
+			if err := syscall.Flock(int(lockFile.Fd()), syscall.LOCK_UN); err != nil {
+				t.Fatalf("flock unlock: %v", err)
+			}
+		}
+		if err := lockFile.Close(); err != nil {
+			t.Fatalf("close audit lock file: %v", err)
+		}
+	}()
+	if err := syscall.Flock(int(lockFile.Fd()), syscall.LOCK_EX); err != nil {
+		t.Fatalf("flock exclusive: %v", err)
+	}
+	locked = true
+
+	close(release)
+	time.Sleep(50 * time.Millisecond)
+	current, ok := runner.GetRun(run.ID)
+	if !ok {
+		t.Fatalf("GetRun(%q): not found", run.ID)
+	}
+	if current.Status == RunStatusCompleted {
+		t.Fatalf("run reached completed status before terminal audit entry could be persisted")
+	}
+
+	if err := syscall.Flock(int(lockFile.Fd()), syscall.LOCK_UN); err != nil {
+		t.Fatalf("flock unlock: %v", err)
+	}
+	locked = false
+	if err := lockFile.Close(); err != nil {
+		t.Fatalf("close audit lock file: %v", err)
+	}
+	lockFile = nil
+
+	waitForStatus(t, runner, run.ID, RunStatusCompleted)
+
+	entries := readAuditEntries(t, auditPath)
+	if got := entries[len(entries)-1].EventType; got != string(EventRunCompleted) {
+		t.Fatalf("last entry = %q, want %q", got, EventRunCompleted)
 	}
 }
 

